[WIP] metric tag autocompletion

alexkorotkikh opened the pull request:
to improve wilcard queries by key performance

alexkorotkikh committed:
metric tag autocompletion

vetinari commented on src/main/java/org/kairosdb/core/http/rest/QueryAutocompleter.java:
```diff
@@ -0,0 +1,49 @@
+        for (final String key : keys) {
+            try {
+                final String metric = extractMetricName(key);
+                if (isWildcard(metric)) {
```
should usually not happen, but this might blow up with an NPE when `extractMetricName()`  returns `null`

alexkorotkikh replied:
I can't imagine a situation when result of `extracMetricName` would be `null`. AFAIK `String::split` returns an array where every object is not null.

rajatparida86 replied:
'extracMetricName' returns null when inout parameter 'key' is empty or null

alexkorotkikh replied:
lol, I'm dumb 🤦‍♂️thanks

alexkorotkikh committed:
Merge branch 'zalando-kairosdb' into autocomplete-metric

alexkorotkikh committed:
fix binding

rajatparida86 commented on src/main/java/org/kairosdb/core/http/rest/QueryAutocompleter.java:
```diff
@@ -0,0 +1,49 @@
+        if (null == key || "".equals(key)) return null;
+        final String[] keyParts = key.split("\\.");
+        final String metricName = keyParts[keyParts.length - 1];
+        return "".equals(metricName) ? keyParts[keyParts.length - 2] : metricName;
```
why `keyParts[keyParts.length - 2]`

alexkorotkikh replied:
This piece of logic was copied [from data-service](https://github.com/zalando-zmon/zmon-data-service/blob/master/src/main/java/de/zalando/zmon/dataservice/data/KairosDBStore.java#L103-L111). I assume there are some (rather rare) cases when key ends with `.`, and then `keyParts.length - 1` is an empty string and and we should use the second from the end section as a `metric`.

alex-on-java replied:
Just FYI: Java's `String.split(regex)` ignores trailing empty strings.
So `aaa.bbb.` will be split to `["aaa", "bbb"]`

alexkorotkikh committed:
check metric for null

alexkorotkikh committed:
check metric for null

rajatparida86 commented on src/main/java/org/kairosdb/core/http/rest/QueryAutocompleter.java:
```diff
@@ -0,0 +1,49 @@
+        completeMetricTag(query);
+    }
+
+    private void completeMetricTag(QueryMetric query) {
```
Two points which should be included in the implementation:
1. We should not add more `metric` tags if the query already contains a `metric` tag.
2. I think we should not do any of the auto-completion if the `key` is not wildcard
I don't see the above 2 points implemented. @alexkorotkikh What do you think?

alexkorotkikh replied:
1. Good catch, thanks! I've added this check to implementation
2. I think we decided to add metric even if key is not a wildcard. It makes an implementation a bit more straightforward because of fewer conditions, while not changing its semantic (metric tag has lower priority than key tag in split index selection, and will be ignored if key is not a wildcard)

alexkorotkikh committed:
check if there's already a metric tag in the query

alex-on-java commented on src/main/java/org/kairosdb/core/http/rest/QueryAutocompleter.java:
```diff
@@ -0,0 +1,49 @@
+    private static final Logger logger = LoggerFactory.getLogger(QueryAutocompleter.class);
+
+    public void complete(QueryMetric query) {
+        completeMetricTag(query);
```
I'm not sure I'm fully understand the reason behind extracting method `completeMetricTag`, instead of inlining it?

alexkorotkikh replied:
The idea is that in future there might be other similar autocompletions added, and then one public method will look like:
```java
    public void complete(QueryMetric query) {
        completeMetricTag(query);
        completeAnotherTag(query);
        completeThridTag(query);
```

alex-on-java commented on src/main/java/org/kairosdb/core/http/rest/QueryAutocompleter.java:
```diff
@@ -0,0 +1,49 @@
+        completeMetricTag(query);
+    }
+
+    private void completeMetricTag(QueryMetric query) {
```
We discussed this with @rajatparida86, sorry that it was not written carefully.
The idea was to autocomplete the query only in situation, when 
* the key has wildcard
* there are no other non-wildcard tags, that could be used as a split index.

The first is obvious, so let me elaborate on the second one.
The `metric` has been added with the second priority (after `key`) in split-index tags. That means, that once we will add the metric tag to the query, it would be used on the first place. 
For some queries it could lead to performance degradation, since the cardinality of `metric` tag in general lower, that the cardinality of `entity`, for example. The same could work for `application_id` and other tags. 

So I would propose to check the query, and add the `metric` only in case when it should definitely improve the performance. 

We, unfortunately, have not so much time to deploy fixes, so from my perspective it will be better to be on a safe side.

alexkorotkikh replied:
I thought that metric would be added as a last-priority index to not mix up cardinalities.

alex-on-java replied:
Yes, that was the part that I missed. We discussed this, and will follow the current approach

alex-on-java commented on src/main/java/org/kairosdb/core/http/rest/QueryAutocompleter.java:
```diff
@@ -0,0 +1,49 @@
+        final Set<String> keys = tags.get("key");
+        final Set<String> metrics = new HashSet<>();
+        for (final String key : keys) {
+            try {
```
As I understood, the exception could be raised only inside the `extractMetricName` method. What do you think to put the `try/catch` block inside that method?

alexkorotkikh replied:
We have to stop processing of the whole `completeMetricTag` method, and that's why we need exception handling on this level

alex-on-java replied:
It totally make sense, but when the `null` is returned, you are also quitting the method. So it will be just one `return;` instead of two. But, as I said, it's not a big deal.

alexkorotkikh replied:
I see your point. Agree, probably in terms of readability it would be better to move try/catch inside the method and return null in case of exception. NB: please make sure that exception is logged

alex-on-java reviewed (commented):
Good job! Just one thing is important from my perspective, the rest comments are optional.

rajatparida86 committed:
Temporary: Only use autocompletion if query start is after metric tag was added to write path

alex-on-java committed:
add tests

rajatparida86 committed:
troubleshooting logs

rajatparida86 committed:
remove troubleshooting logs

rajatparida86 committed:
repair test case

alexkorotkikh commented on src/main/java/org/kairosdb/core/http/rest/MetricsResource.java:
```diff
@@ -126,10 +127,15 @@
-	public MetricsResource(KairosDatastore datastore, QueryParser queryParser,
-						   KairosDataPointFactory dataPointFactory, QueryMeasurementProvider queryMeasurementProvider, Tracer tracer) {
+	public MetricsResource(KairosDatastore datastore,
+						   QueryParser queryParser,
```
Please align indentation.

alexkorotkikh reviewed (commented):
2 minor comments, other than that :+1:

alex-on-java committed:
fix indentation

alex-on-java committed:
move exception handling inside the extraction method

we catch potential exception in place
where there is a possibility it could be raised

alex-on-java committed:
set exact time, when we can safely remove the time check

alex-on-java commented:
👍

rajatparida86 commented:
👍
